dags/utils/knack.py

```diff
@@ -20,7 +20,19 @@ def get_date_filter_arg(should_replace_monthly=False, **context):
         Str or None: the -d flag and ISO date string or None
     """
     today = now()
-    prev_start_date = context.get("prev_start_date_success") or today.isoformat()
+    prev_start_date = context.get("prev_start_date_success")
+
+    if isinstance(prev_start_date, DateTime):
+        prev_start_date = prev_start_date.isoformat()
+    elif prev_start_date:
+        try:
+            prev_start_date = parse(prev_start_date).isoformat()
+        except ValueError:
+            prev_start_date = today.isoformat()
```

Comment:
@frankhereford i'm a bit confused about this `elif` block. the Airflow [template string docs](https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html) suggest that `prev_start_date_success` is either going to be a `DateTime` instance or `None`, but this would suggest it may also sometimes be a string?

Reply:
ah, is it that the context variable is always a string, and that string is no longer in ISO format? e.g. airflow broke the shape of their own context vars by upgrading pendulum? if that's the case—yikes.

either way, its seems like there must be part of this if block that will never execute 🤷

Reply:
Yea, that's a great question John. The reason for these hoops is because the date we were getting out of the `context.get()` function was a string. We had previous dates coming out that we had to handle in addition to the form of new ones we were creating.

Reply:
Airflow handled the transition correctly, I believe, but relied on the stack to be upgraded in such a way that a specific minor version not be skipped entirely, which we did. I'm not saying they couldn't have left more handlers in place for longer or forever, but I can see that they didn't want to support something they said they were no longer supporting. They were very clear that they were intent on moving to Pendulum 3, and we just missed the whole thing by doing such a big upgrade.

Reply:
@johnclary wrote: 

> either way, its seems like there must be part of this if block that will never execute 🤷

Forgive me if I'm missing something here, but I read this as:

If we get a `DateTime` out of the context, use it
Else, if we get something out of the context, try to parse it as a string into a DateTime, and if that fails, just use now. 
And finally, if we get nothing out of the context, use now.

Can you help me understand which part is unreachable? I'd be happy to sync with you if that's easier.

Reply:
my confusion is why `context.get("prev_start_date_success")` may sometimes return a string and sometimes a `DateTime` instance. the previous code assumed this value was always a string or `None`.  So the new version of the code would be like this:
```python
if prev_start_date:
    prev_start_date = parse(prev_start_date).isoformat()
else:
    prev_start_date = today.isoformat()
```

Reply:
Ah, that makes so much sense!

The reason was -- just for the immediate time while we were in the transition, we had contexts from the 2.6 era which were coming out of the context as a string, as they had been stored as a string, and we were storing in DataTimes, which the next time the dag ran, we'd have DateTimes coming out. This "handle them both" fix was for the transition, since we missed the versions where Airflow had this "handle them both" fix internal to it.

Reply:
If I knew that all of our DAGs which have ever run had run again since the upgrade, this code could be simplified to eliminate one of the branches of the condition. However, I don't know that for sure. We could write up an issue to find out and to make that enhancement if you think it'd be a good thing to do.

Reply:
oh, the context API stores data from earlier airflow versions! that's bananas 😵 

would you mind adding a comment to the block that says something about backwards compatible support for string instances of this context var?

i'm still pretty baffled, because the release notes for [this change](https://airflow.apache.org/docs/apache-airflow/stable/release_notes.html#context-variables-prev-execution-date-success-and-prev-execution-date-success-are-now-pendulum-datetime) are under the Airflow v2 release notes.

> Context variables prev_execution_date_success and prev_execution_date_success are now pendulum.DateTime


🤔

Reply:
hmm, i guess i'm not convinced now. are you sure [this block](https://github.com/cityofaustin/atd-airflow/pull/231/files#r1610348924) doesn't get the job done?

Reply:
Sure thing on the comment. 

Yea, I do hear you on the baffling observation you made there. I should point out that this code was born out of a trial-and-error debugging session because we had DAGs failing. I'm not sure if it's Airflow misbehaving or maybe our own code misbehaving -- no idea what the root of the strings are -- but I was seeing strings and DateTimes and this seemed reasonable to handle both.

Reply:
@johnclary and I had an opportunity to discuss this in a huddle, and we're both interested in finding out the real details of what happened here. I think he has a fuller picture of the timeline of events after reading documentation. In terms of what we're going to do right now, I'm going to put the simplified block of code which he suggests in place, and if it works, then I'm going to revise this PR to reflect it. Thanks again to @johnclary for the discussion and the very valuable development perspective of valuing a full understanding and not acting blindly.